e2e_playwright/st_toast_test.py

```diff
@@ -0,0 +1,74 @@
+# Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022-2024)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from playwright.sync_api import Page, expect
+
+from e2e_playwright.conftest import ImageCompareFunction
+
+
+def test_default_toast_rendering(
+    themed_app: Page, assert_snapshot: ImageCompareFunction
+):
+    """Test that toasts are correctly rendered."""
+    themed_app.keyboard.press("r")
+    toasts = themed_app.get_by_test_id("stToast")
+    expect(toasts).to_have_count(2)
+    toasts.nth(1).hover()
+
+    expect(toasts.nth(1)).to_have_text("🐶 This is a default toast messageClose")
+    assert_snapshot(toasts.nth(1), name="toast-default")
+
+
+def test_collapsed_toast_rendering(
+    themed_app: Page, assert_snapshot: ImageCompareFunction
+):
+    """Test collapsed long toasts are correctly rendered."""
+    themed_app.keyboard.press("r")
+    toasts = themed_app.get_by_test_id("stToast")
+    expect(toasts).to_have_count(2)
+    toasts.nth(0).hover()
+
+    expect(toasts.nth(0)).to_have_text(
+        "🦄 Random toast message that is a really really really really really really really long message, going wayview moreClose"
+    )
+    assert_snapshot(toasts.nth(0), name="toast-collapsed")
+
+
+def test_expanded_toast_rendering(
+    themed_app: Page, assert_snapshot: ImageCompareFunction
+):
+    """Test expanded long toasts are correctly rendered."""
+    themed_app.keyboard.press("r")
+    toasts = themed_app.get_by_test_id("stToast")
+    expect(toasts).to_have_count(2)
+    toasts.nth(0).hover()
+
+    themed_app.get_by_text("view more").first.click()
+
+    expect(toasts.nth(0)).to_have_text(
```

Comment:
This test still seems to be slightly flaky. From looking at the video, one guess might be that playwright is a bit to fast here and is actually trying to click on the `"view more"` of the previous toast (pre-rerun). One hacky solution to make it more stable might be to add a short timeout after the `press("r")`:
```
themed_app.keyboard.press("r")
themed_app.wait_for_timeout(250)
```

Or a probable better solution is to wait for the two toasts to disappear before pressing `r`:

``` 
toasts = themed_app.get_by_test_id("stToast")
expect(toasts).to_have_count(0)
themed_app.keyboard.press("r")
```

We can probably reuse the same solution for the other toast tests to lower the risk of flakiness.

Reply:
Ohh okay that makes a lot of sense - thanks for the thorough review 👍🏼

Reply:
Argh, the default 5-second waiting timeout doesn't seem to be enough :(  Another idea: maybe just trigger the toasts on a button click. That gives a bit more control over triggering the toasts.